### PR TITLE
[Admin] - User Edit Page - The wrong user is updated

### PR DESCRIPTION
**Pull request by popsorin**

### CODE REVIEW

File: public/views/assets/admin-user-details.phtml [hunk 1]
```diff
@@ -49,9 +49,11 @@
...
         </div>
 
-        <?php $displayName = (isset($name) && $name !== "") ? $name : "" ?>
-        <?php $displayEmail = (isset($email) && $email !== "") ? $email : "" ?>
-        <?php $errorMessage = (isset($errorMessage) && $errorMessage !== "") ? $errorMessage : "" ?>
+        <?php $displayName = ($name) ?? ""?>
+        <?php $displayEmail = ($email) ?? "" ?>
+        <?php $errorMessage = ($errorMessage) ?? "" ?>
+        <?php $userRole =  (isset($userRole) && $userRole === "admin") ? "admin" : "candidate"  ?>
+        <?php $otherRole =  (isset($userRole) && $userRole !== "admin") ? "admin" : "candidate"  ?>
```
Comment (Raresmldvn) [inline public/views/assets/admin-user-details.phtml:56]: first condition is useless (if it is not set, it is null so you can remove it), and I can't possibly imagine what "other role is"

File: public/views/assets/admin-question-details.phtml [hunk 2]
```diff
@@ -48,42 +48,18 @@
         <div class="text-center">
             <h1>New question / Edit question</h1>
         </div>
```
Comment (victortodoran) [inline public/views/assets/admin-question-details.phtml:50]: IMO it should not be the responsibility of the template to handle the state of the $question. Even more so I would expect we work with object not scalar type and such $question be an object.

File: public/views/assets/admin-user-details.phtml [hunk 3]
```diff
@@ -49,9 +49,11 @@
             <h1>New user / Edit user</h1>
         </div>
```
Comment (victortodoran) [inline public/views/assets/admin-user-details.phtml:51]: IMO you have way too much logic in your template. The role of the template is to display stuff. Not check and assign values.

File: src/Controller/AdminController.php [hunk 4]
```diff
@@ -39,7 +39,7 @@ public function __construct(
     public function showDashboard(Request $request, array $attributes): Response
     {
```
Comment (victortodoran) [inline src/Controller/AdminController.php:40]: In the future consider adding a wrapper over the session as explained by Andra roughly a month ago. Also I encourage you to start working with objects in the detriment of scalary types.

File: src/Controller/LoginController.php [hunk 5]
```diff
@@ -75,12 +75,12 @@ public function login(Request $request, array $attributes): Response
     {
         $this->session->start();
         try {
-            $entity = $this->service->login(["name" => $request->getParameter("name"), "password" => $request->getParameter("password")]);
+            $entity = $this->service->login(["email" => $request->getParameter("email"), "password" => $request->getParameter("password")]);
         }
         catch (WrongPasswordException $exception) {
             return $this->createResponse($request, "301", "Location", ["/"]);
         }
```
Comment (victortodoran) [inline src/Controller/LoginController.php:82]: Please make an experiment and get back to me. What if you set an obect in the session. Will you have at the next request?

File: src/Controller/QuestionTemplateController.php [hunk 6]
```diff
@@ -28,26 +30,38 @@ class QuestionTemplateController extends AbstractController
     /**
      * @var SessionInterface
      */
-    protected $session;
+    private $session;
 
     /**
      * @var QuestionTemplateService
      */
```
Comment (victortodoran) [inline src/Controller/QuestionTemplateController.php:37]: What service? Please use a more descriptive name for this variable. Browsing through code I should be able to know the variable type from its name, ideally.

File: src/Controller/UserController.php [hunk 7]
```diff
@@ -20,6 +21,8 @@ class UserController extends AbstractController
 
     const ADMIN_USER_LISTING_PAGE = "admin-users-listing.phtml";
```
Comment (victortodoran) [inline src/Controller/UserController.php:23]: EXCEPTION_PAGE_TEMPLATE

File: src/Controller/UserController.php [hunk 8]
```diff
@@ -35,24 +38,32 @@ class UserController extends AbstractController
      */
     private $factory;
```
Comment (victortodoran) [inline src/Controller/UserController.php:40]: userValidator if it is a specific validator

File: src/Controller/UserController.php [hunk 9]
```diff
@@ -35,24 +38,32 @@ class UserController extends AbstractController
...
+     * @param EntityValidatorInterface $validator
      */
     public function __construct(
         RendererInterface $renderer,
         UserService $service,
         SessionInterface $session,
-        UserFactory $factory
+        UserFactory $factory,
+        EntityValidatorInterface $validator
     )
```
Comment (victortodoran) [inline src/Controller/UserController.php:60]: replace
)
{
with
) {

File: src/Controller/UserController.php [hunk 10]
```diff
@@ -62,30 +73,72 @@ public function __construct(
...
-        $entity = $this->factory->createFromRequest($request, "name", "email", "password", "role");
+        $user = $this->factory->createFromRequest($request, "name", "email", "password", "role");
+
         try {
-            $this->service->add($entity);
-        } catch (UserAlreadyExistsException $exception) {
+            $this->validator->validate($user);
+        } catch (InvalidUserException $exception) {
+            $message = "";
+
```
Comment (victortodoran) [inline src/Controller/UserController.php:82]: Extract error message building logic in separate method. Are u sure your not showing some sensitive information tot the end user with this logic.

File: src/Controller/UserController.php [hunk 11]
```diff
@@ -62,30 +73,72 @@ public function __construct(
...
+            while ($exception->getPrevious() !== null) {
+                $message .= $exception->getMessage() . '<br>';
+                $exception = $exception->getPrevious();
+            }
+
```
Comment (victortodoran) [inline src/Controller/UserController.php:87]: Why not pass the $user entity directly to the template and let it figure out what to show there. Please refactor this.

File: src/Controller/UserController.php [hunk 12]
```diff
@@ -62,30 +73,72 @@ public function __construct(
...
 
     /**
+     *
+     *
      * @param Request $request
      * @param array $attributes
      * @return Response
      */
     public function update(Request $request, array $attributes): Response
     {
```
Comment (victortodoran) [inline src/Controller/UserController.php:112]: What if attribute is not set? You will get an PHP error. Why not add a wrapper over this array. An object that has the responsibility of returning parameter values. The request object could also be responsible of retrieving parameter values.

File: src/Controller/UserController.php [hunk 13]
```diff
@@ -62,30 +73,72 @@ public function __construct(
...
-        $this->service->update($entity);
+        $id = $attributes["id"];
+        $updatedUser = $this->factory->createFromRequest($request, "name", "email", "password","role");
+        $updatedUser->setId($id);
+
+        try {
+            $this->validator->validate($updatedUser);
+        } catch (InvalidUserException $exception) {
+            $message = "";
+
```
Comment (victortodoran) [inline src/Controller/UserController.php:121]: Same story about building error messages here as above.

File: src/Controller/UserController.php [hunk 14]
```diff
@@ -118,27 +171,29 @@ public function getAll(Request $request, array $attributes): Response
...
      * @return Response
-     * Returns the page for the add functionality
+     * Returns the page for the update functionality
      */
-    public function getUserDetails(Request $request, array $attributes): Response
+    public function showEditUserPage(Request $request, array $attributes): Response
     {
-        $user = $this->service->getUserDetails($attributes["id"]);
         $this->session->start();
+        $user = $this->service->findUserById($attributes["id"]);
```
Comment (victortodoran) [inline src/Controller/UserController.php:180]: Pass the $user object to the template.

File: src/Factory/AnswerChoiceInstanceFactory.php [hunk 15]
```diff
@@ -22,7 +22,8 @@ public function createFromRequest(
         string $questionInstanceIdKey,
         string $isSelectedKey,
         string $isCorrectKey
```
Comment (victortodoran) [inline src/Factory/AnswerChoiceInstanceFactory.php:24]: replace with ): SomeType {

File: src/Factory/AnswerTextInstanceFactory.php [hunk 16]
```diff
@@ -19,7 +19,8 @@ public function createFromRequest(
         Request $request,
         string $questionInstanceIdKey,
         string $textKey
```
Comment (victortodoran) [inline src/Factory/AnswerTextInstanceFactory.php:21]: https://www.php-fig.org/psr/psr-12/

File: src/Factory/UserFactory.php [hunk 17]
```diff
@@ -23,7 +23,8 @@ public function createFromRequest(
         string $emailKey,
         string $passwordKey,
         string $roleKey
```
Comment (victortodoran) [inline src/Factory/UserFactory.php:25]: https://www.php-fig.org/psr/psr-12/

File: src/Service/Validator/UserValidator.php [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace Quiz\Service\Validator;
+
+
+use Quiz\Service\Exception\InvalidUserException;
+use ReallyOrm\Entity\EntityInterface;
+use ReallyOrm\Repository\RepositoryInterface;
+
+class UserValidator implements EntityValidatorInterface
+{
```
Comment (victortodoran) [inline src/Service/Validator/UserValidator.php:12]: Does the name have to be exactly 3 chars? And the password 8? If not please replace with more descriptive constant names.

File: src/Service/Validator/UserValidator.php [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     *
+     * Validates the given user for the add functionality
+     * The user email is unique and must not be taken by somebody else
+     *
+     * @param EntityInterface $user
+     * @return InvalidUserException|null
+     */
```
Comment (victortodoran) [inline src/Service/Validator/UserValidator.php:69]: return type.

File: src/Service/Validator/UserValidator.php [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     *
+     * Validates the given user for the update functionality
+     * The user email is unique and must not be taken by somebody else
+     *
+     * @param EntityInterface $user
+     * @return InvalidUserException|null
+     */
```
Comment (victortodoran) [inline src/Service/Validator/UserValidator.php:89]: return type

File: src/Service/Validator/UserValidator.php [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+     * Verify if the user can be saved in the database.
+     *
+     * Throws an exception containing every issue found with the user data
+     *
+     * @param EntityInterface $user
+     * @return void
+     * @throws InvalidUserException
+     */
+    public function validate(EntityInterface $user): void
+    {
```
Comment (victortodoran) [inline src/Service/Validator/UserValidator.php:41]: Which design pattern could you implement to get rid of all the ifs in this method?

**Pull request merged**